src/main/java/build/buildfarm/worker/shard/Worker.java

```diff
@@ -100,6 +117,118 @@
   private final ShardBackplane backplane;
   private final LoadingCache<String, Instance> workerStubs;
 
+  class LocalCasWriter implements CasWriter {
+    public void write(Digest digest, Path file) throws IOException, InterruptedException {
+      insertFileLocally(digest, file);
+    }
+
+    private void insertFileLocally(Digest digest, Path file)
+        throws IOException, InterruptedException {
+
+      Write write = getLocalWrite(digest);
+
+      try (OutputStream out =
+              write.getOutput(/* deadlineAfter=*/ 1, /* deadlineAfterUnits=*/ DAYS, () -> {});
+          InputStream in = Files.newInputStream(file)) {
+        ByteStreams.copy(in, out);
+      } catch (IOException e) {
+        if (!write.isComplete()) {
+          write.reset(); // we will not attempt retry with current behavior, abandon progress
+          if (e.getCause() != null) {
+            Throwables.propagateIfInstanceOf(e.getCause(), InterruptedException.class);
+          }
+          throw e;
+        }
+      }
+    }
```

Comment:
The replication is strong with this one. We already have an input stream supplier iirc - reduce the duplicity between this and the blob insert with that